sphinx/ext/intersphinx.py

```diff
@@ -533,17 +532,122 @@ def run(self) -> tuple[list[Node], list[system_message]]:
         assert self.name == self.orig_name.lower()
         inventory, name_suffix = self.get_inventory_and_name_suffix(self.orig_name)
         if inventory and not inventory_exists(self.env, inventory):
-            logger.warning(__('inventory for external cross-reference not found: %s'),
-                           inventory, location=(self.env.docname, self.lineno))
+            logger.warning(
+                __('inventory for external cross-reference not found: %s'),
+                inventory,
+                type='intersphinx',
+                subtype='external',
+                location=(self.env.docname, self.lineno),
+            )
             return [], []
 
-        role_name = self.get_role_name(name_suffix)
+        domain_name, role_name = self._get_domain_role(name_suffix)
+
         if role_name is None:
-            logger.warning(__('role for external cross-reference not found: %s'), name_suffix,
-                           location=(self.env.docname, self.lineno))
+            logger.warning(
+                __('external cross-reference suffix not valid: %s'),
+                name_suffix,
+                type='intersphinx',
+                subtype='external',
+                location=(self.env.docname, self.lineno),
+            )
             return [], []
 
-        result, messages = self.invoke_role(role_name)
+        # attempt to find a matching role function
+        role_func: None | RoleFunction
+
+        if domain_name is not None:
+            # the user specified a domain, so we only check that
+            if (domain := self.env.domains.get(domain_name)) is None:
+                logger.warning(
+                    __('domain for external cross-reference not found: %s'),
+                    domain_name,
+                    type='intersphinx',
+                    subtype='external',
+                    location=(self.env.docname, self.lineno),
+                )
+                return [], []
+            if (role_func := domain.roles.get(role_name)) is None:
+                msg = 'role for external cross-reference not found in domain %s: %s'
+                if (
+                    object_types := domain.object_types.get(role_name)
+                ) is not None and object_types.roles:
+                    logger.warning(
```

Comment:
I suggest having a dedicated method for adding `type` and `subtype` + `location`. The reason is that you could have less lines of codes.

Reply:
yep fair

Reply:
done